[mypyc] Support separate compilation

msullivan opened the pull request:
This adds support for separate compilation to mypyc.

A `separate` argument is added to `mypycify`. If `True`, mypyc places every
module its own separate shared library. Otherwise it can take a list of groups
of source files that should be placed in a shared library together.

The shared libraries communication with each other using the C API's intended
mechanism for communicating between C extension modules: Capsules.
Each library creates a table of pointers to all of its exported data and stores a pointer
to it in a capsule stored as a module attribute. When a library is loaded, it
loads the linking table capsules from all its dependencies and copies them
into a local copy of the table. (To eliminate the need for a pointer indirection when
accessing it.)

This adds a test mode that will run all multi-module run tests in separate compilation mode.
I also manually tested mypy itself compiled in separate compilation mode.

This supports a limited form of incremental compilation already: only modules
that changed or had a header they depend on change will be recompiled by the
C compiler. The entire project still needs to go through the entire mypy/mypyc
front-end and middle-end, however. I expect to have a PR that adds support for 
hooking into incremental mode next week.

This is progress on mypyc/mypyc#682.

msullivan commented:
(I haven't squashed the commit history yet because it was useful for me in revising this PR and might be useful to me in addressing feedback. But it probably isn't useful to anybody reviewing this, because I made some changes in early commits that I undid in later commits and because github bizarrely insists on sorting commits by their timestamp.)

JukkaL commented on mypyc/build.py:
```diff
@@ -160,14 +161,13 @@ def include_dir() -> str:
+               ) -> Tuple[List[List[Tuple[str, str]]], str]:
     """Drive the actual core compilation step.
 
     Returns the C source code and (for debugging) the pretty printed IR.
```
Document `groups`.

JukkaL commented on mypyc/build.py:
```diff
@@ -276,13 +283,57 @@ def write_file(path: str, contents: str) -> None:
+    sources: List[BuildSource],
+    separate: Union[bool, List[Tuple[List[str], Optional[str]]]],
+    use_shared_lib: bool,
+) -> List[Tuple[List[BuildSource], Optional[str]]]:
```
Add docstring.

JukkaL commented on mypyc/build.py:
```diff
@@ -276,13 +283,57 @@ def write_file(path: str, contents: str) -> None:
+        (group,
+         name or (
+             shared_lib_name([source.module for source in group]) if use_shared_lib else None))
+        for group, name in groups]
```
I found this a bit hard to understand. It might be easier to read if it used for and if statements.

JukkaL commented on mypyc/build.py:
```diff
@@ -276,13 +283,57 @@ def write_file(path: str, contents: str) -> None:
+    return groups
+
+
+def get_header_deps(cfiles: List[Tuple[str, str]]) -> List[str]:
```
Document the `cfiles` argument.

JukkaL commented on mypyc/build.py:
```diff
@@ -276,13 +283,57 @@ def write_file(path: str, contents: str) -> None:
+    verbose: bool = False,
+    strip_asserts: bool = False,
+    separate: Union[bool, List[Tuple[List[str], Optional[str]]]] = False,
+) -> List[Extension]:
```
Document additional arguments -- at least `separate`.

JukkaL commented on mypyc/emit.py:
```diff
@@ -30,27 +31,36 @@ class HeaderDeclaration:
 
 
 class EmitterContext:
     """Shared emitter state for an entire compilation unit."""
```
Maybe discuss what a compilation unit means in this context?

JukkaL commented on mypyc/emit.py:
```diff
@@ -125,6 +135,28 @@ def new_label(self) -> str:
         self.context.temp_counter += 1
         return '__LL%d' % self.context.temp_counter
 
+    def get_lib_prefix(self, obj: Union[str, ClassIR, FuncDecl],
```
Document what a `str` `obj` argument means. Document `is_variable`.

JukkaL commented on mypyc/emitmodule.py:
```diff
@@ -104,17 +124,26 @@ def encode_bytes_as_c_string(b: bytes) -> Tuple[str, int]:
     return '"{}"'.format(escaped), len(b)
 
 
+def pointerize(s: str, name: str) -> str:
```
Add docstring.

JukkaL commented on mypyc/emitmodule.py:
```diff
@@ -104,17 +124,26 @@ def encode_bytes_as_c_string(b: bytes) -> Tuple[str, int]:
                  modules: List[Tuple[str, ModuleIR]],
                  source_paths: Dict[str, str],
                  shared_lib_name: Optional[str],
+                 group_map: Dict[str, Optional[str]],
```
Document `group_map` (and maybe other arguments).

JukkaL commented on mypyc/emitmodule.py:
```diff
@@ -217,16 +257,44 @@ def generate_c_for_modules(self) -> List[Tuple[str, str]]:
+                                 ''.join(ext_declarations.fragments)),
                                 ]
 
+    def generate_linking_struct(self, decl_emitter: Emitter, code_emitter: Emitter) -> None:
```
Add docstring. Document the linking table somewhere (both the purpose and the structure).

JukkaL commented on mypyc/emitmodule.py:
```diff
@@ -244,34 +312,66 @@ def generate_shared_lib_init(self, emitter: Emitter) -> None:
+            'if (!capsule) {',
+            'goto fail;',
+            '}',
+            'res = PyObject_SetAttrString(module, "exports", capsule);',
```
Does this get added to the module namespace? What if there's an attribute with name `exports`?

msullivan replied:
This is part of the module namespace for the shared library module, not the actual module.

JukkaL commented on mypyc/emitmodule.py:
```diff
@@ -244,34 +312,66 @@ def generate_shared_lib_init(self, emitter: Emitter) -> None:
             '',
         )
 
+        emitter.emit_lines(
```
Document the whole capsule business in detail somewhere.

JukkaL commented on mypyc/test/test_run.py:
```diff
@@ -251,3 +245,13 @@ class TestRunMultiFile(TestRun):
+
+
+# Run the main multi-module tests in separate compliation mode
+class TestRunSeparate(TestRun):
```
Would it be useful to add some new test cases specific to separate compilation mode, or are the existing tests good enough?

msullivan replied:
I'm adding two cases I remember hitting to one of the tests, but otherwise they were pretty good.

It is probably worth adding testing of manually specifying group assignments, but that will require some infrastructure so I'll leave it to a follow-up diff I think. (I've tested that manually.)

JukkaL commented on mypyc/emitmodule.py:
```diff
@@ -76,16 +87,25 @@ def compile_modules_to_c(result: BuildResult, module_names: List[str],
-                                compiler_options.multi_file)
-    return generator.generate_c_for_modules()
+
+    # Generate all the groups
```
Explain this in a bit more detail?

JukkaL commented on mypyc/emitmodule.py:
```diff
@@ -42,16 +42,27 @@ def parse_and_typecheck(sources: List[BuildSource], options: Options,
+    result: BuildResult,
+    compiler_options: CompilerOptions,
+    errors: Errors,
+    groups: List[Tuple[List[BuildSource], Optional[str]]],
```
Document the `groups` argument (and maybe some other arguments as well).

JukkaL commented on mypyc/emitmodule.py:
```diff
@@ -216,62 +304,151 @@ def generate_c_for_modules(self) -> List[Tuple[str, str]]:
+        the need for a pointer indirection when accessing it).
+
+        Then, all calls to functions in another group and accesses to statics
+        from another group are done indirectly via the export table.
```
Maybe give a concrete example of what the export table looks like, and another group can access and item from it (using C fragments)?

JukkaL commented on mypyc/emitmodule.py:
```diff
@@ -216,62 +304,151 @@ def generate_c_for_modules(self) -> List[Tuple[str, str]]:
 
-        A shared library contains all of the actual code for a set of modules.
+        A shared library contains all of the actual code for a
+        compilation group.
```
A more concrete explanation would be helpful. For each module in a group, we generate a shim with a name derived from the module name (?) and a shared library with a name that is not derived in a straightforward way from the module name, right? But if the group only has one module, do we need a separate shared lib?

msullivan replied:
We need the separate shared lib for an annoying and perhaps somewhat subtle reason: the reading of export tables from dependencies is done as a separate phase from actually executing module top levels, because I wanted to make sure that the export tables were always initialized and couldn't get mixed up trouble due to import cycles.

But:
 1. I think I have figured out a way to keep that property while allowing just one extension module per python module, though it is somewhat unpleasant
 2. Maybe it isn't actually important anyway

I'll document what's going on here a bit more and consider one of those alternatives in a follow-up maybe

msullivan commented on mypyc/namegen.py:
```diff
@@ -36,18 +36,22 @@ class NameGenerator:
     though not very usable.
+
+    FIXME: Currently when performing seperate compilation we just
+    totally give up on shortening names.
```
Mostly just a little more verbose.

msullivan commented on mypyc/emit.py:
```diff
@@ -126,6 +136,30 @@ def new_label(self) -> str:
+        target_group_name = groups.get(module_name)
+        if target_group_name and target_group_name != self.context.group_name:
+            self.context.group_deps.add(target_group_name)
+            return 'exports_{}.'.format(target_group_name)
```
Yeah, I will change the names to be based on the real names for the auto-generated single-module groups.

msullivan replied:
Aha, and if I do that and also fix NameGenerator to work at a group level, I think everything works out pretty nicely here.

msullivan commented:
OK I think I have addressed all the feedback. There was also some function renaming.

One major pair of substantive changes went in, which is to generate group names from single-module groups in a legible manner and to restore namegen module name shortening. This depends on https://github.com/python/mypy/pull/7676 now, though.

JukkaL commented on mypyc/emitmodule.py:
```diff
@@ -73,17 +131,32 @@ def compile_modules_to_c(result: BuildResult, module_names: List[str],
+
+    # Generate C code for each compilation group. Each group will be
+    # compiled into a separate extension module, so we produce
+    # a separate group of C source for each of them.
```
"Group of C source" is not entirely clear. Do we generate a separate file for each file in a group?

JukkaL commented on mypyc/emitmodule.py:
```diff
@@ -104,32 +177,62 @@ def encode_bytes_as_c_string(b: bytes) -> Tuple[str, int]:
 
-class ModuleGenerator:
+def pointerize(s: str, name: str) -> str:
+    """Given C type and a variable name, return a declaration of a pointer to it."""
```
I'm not sure if the docstring matches the implementation. It looks like `s` should have both a type and a variable name, since we are replacing `name` in it. Also, is it possible that there will be multiple matches to replace? I assume that the variable name needs to be sufficiently unique.

JukkaL commented on mypyc/emitmodule.py:
```diff
@@ -104,32 +177,62 @@ def encode_bytes_as_c_string(b: bytes) -> Tuple[str, int]:
                  multi_file: bool) -> None:
+        """Generator for C source for a compilation group.
+
+        We do this by just regexping the source, which is a bit simpler than
```
Explain what the generated C source includes. This can be understood (based on just this docstring) as being some grepped subset of some other source. Also this would be clearer if the distinction between the "source source" and generated source was more explicit. Maybe a few more sentences would help here?

msullivan replied:
Oh lurr that regexping comment got copy-pasted from elsewhere and is just wrong

JukkaL commented:
Oh and there is a build failure.
